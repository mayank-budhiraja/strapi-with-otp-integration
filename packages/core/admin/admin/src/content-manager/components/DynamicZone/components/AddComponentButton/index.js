/**
 *
 * AddComponentButton
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import styled from 'styled-components';
import PlusCircle from '@strapi/icons/PlusCircle';
import { BaseButton } from '@strapi/design-system/BaseButton';
import { Box } from '@strapi/design-system/Box';
import { Flex } from '@strapi/design-system/Flex';
import { Typography } from '@strapi/design-system/Typography';
import { getTrad } from '../../../../utils';

const StyledAddIcon = styled(PlusCircle)`
  transform: ${({ $isOpen }) => ($isOpen ? 'rotate(45deg)' : 'rotate(0deg)')};
  > circle {
    fill: ${({ theme, $hasError }) =>
      $hasError ? theme.colors.danger200 : theme.colors.neutral150};
  }
  > path {
    fill: ${({ theme, $hasError }) =>
      $hasError ? theme.colors.danger600 : theme.colors.neutral600};
  }
`;

const StyledButton = styled(BaseButton)`
  border-radius: 26px;
  border-color: ${({ theme }) => theme.colors.neutral150};
  background: ${({ theme }) => theme.colors.neutral0};
  padding-top: ${({ theme }) => theme.spaces[3]};
  padding-right: ${({ theme }) => theme.spaces[4]};
  padding-bottom: ${({ theme }) => theme.spaces[3]};
  padding-left: ${({ theme }) => theme.spaces[4]};

  box-shadow: ${({ theme }) => theme.shadows.filterShadow};

  svg {
    height: ${({ theme }) => theme.spaces[6]};
    width: ${({ theme }) => theme.spaces[6]};
    > path {
      fill: ${({ theme }) => theme.colors.neutral600};
    }
  }
  &:hover {
    color: ${({ theme }) => theme.colors.primary600} !important;
    ${Typography} {
      color: ${({ theme }) => theme.colors.primary600} !important;
    }

    ${StyledAddIcon} {
      > circle {
        fill: ${({ theme }) => theme.colors.primary600};
      }
      > path {
        fill: ${({ theme }) => theme.colors.neutral100};
      }
    }
  }
  &:active {
    ${Typography} {
      color: ${({ theme }) => theme.colors.primary600};
    }
    ${StyledAddIcon} {
      > circle {
        fill: ${({ theme }) => theme.colors.primary600};
      }
      > path {
        fill: ${({ theme }) => theme.colors.neutral100};
      }
    }
  }
`;

const BoxFullHeight = styled(Box)`
  height: 100%;
`;

const AddComponentButton = ({
  hasError,
  hasMaxError,
  hasMinError,
  isDisabled,
  isOpen,
  label,
  missingComponentNumber,
  name,
  onClick,
}) => {
  const { formatMessage } = useIntl();
  const addLabel = formatMessage(
    {
      id: getTrad('components.DynamicZone.add-component'),
      defaultMessage: 'Add a component to {componentName}',
    },
    { componentName: label || name }
  );
  const closeLabel = formatMessage({ id: 'app.utils.close-label', defaultMessage: 'Close' });
  let buttonLabel = isOpen ? closeLabel : addLabel;

  if (hasMaxError && !isOpen) {
    buttonLabel = formatMessage({
      id: 'components.Input.error.validation.max',
      defaultMessage: 'The value is too high.',
    });
  }

  if (hasMinError && !isOpen) {
    buttonLabel = formatMessage(
      {
        id: getTrad(`components.DynamicZone.missing-components`),
        defaultMessage:
          'There {number, plural, =0 {are # missing components} one {is # missing component} other {are # missing components}}',
      },
      { number: missingComponentNumber }
    );
  }

  return (
    <>
      <Flex justifyContent="center">
        <Box style={{ cursor: isDisabled ? 'not-allowed' : 'pointer' }}>
          <StyledButton type="button" onClick={onClick} disabled={isDisabled} hasError={hasError}>
            <Flex>
              <BoxFullHeight aria-hidden paddingRight={2}>
                <StyledAddIcon $isOpen={isOpen} $hasError={hasError && !isOpen} />
              </BoxFullHeight>
              <Typography
                variant="pi"
                fontWeight="bold"
                textColor={hasError && !isOpen ? 'danger600' : 'neutral500'}
              >
                {buttonLabel}
              </Typography>
            </Flex>
          </StyledButton>
        </Box>
      </Flex>
    </>
  );
};

AddComponentButton.defaultProps = {
  label: '',
  missingComponentNumber: 0,
};

AddComponentButton.propTypes = {
  label: PropTypes.string,
  hasError: PropTypes.bool.isRequired,
  hasMaxError: PropTypes.bool.isRequired,
  hasMinError: PropTypes.bool.isRequired,
  isDisabled: PropTypes.bool.isRequired,
  isOpen: PropTypes.bool.isRequired,
  missingComponentNumber: PropTypes.number,
  name: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default AddComponentButton;
