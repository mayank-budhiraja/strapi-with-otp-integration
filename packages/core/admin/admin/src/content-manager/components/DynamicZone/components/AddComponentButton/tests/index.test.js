/**
 *
 * Tests for AddComponentButton
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider, lightTheme } from '@strapi/design-system';
import { IntlProvider } from 'react-intl';
import AddComponentButton from '../index';

describe('<AddComponentButton />', () => {
  it('renders and matches the snapshot', () => {
    const { container } = render(
      <ThemeProvider theme={lightTheme}>
        <IntlProvider locale="en" messages={{}} defaultLocale="en">
          <AddComponentButton
            label="test"
            isOpen={false}
            name="name"
            onClick={jest.fn()}
            hasError={false}
            hasMaxError={false}
            hasMinError={false}
          />
        </IntlProvider>
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders and matches the snapshot when the isOpen prop is truthy', () => {
    const { container } = render(
      <ThemeProvider theme={lightTheme}>
        <IntlProvider locale="en" messages={{}} defaultLocale="en">
          <AddComponentButton
            label="test"
            isOpen
            name="name"
            onClick={jest.fn()}
            hasError={false}
            hasMaxError={false}
            hasMinError={false}
          />
        </IntlProvider>
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('displays the name of the dz when the label is empty', () => {
    const { container } = render(
      <ThemeProvider theme={lightTheme}>
        <IntlProvider locale="en" messages={{}} defaultLocale="en">
          <AddComponentButton
            label=""
            isOpen={false}
            name="name"
            onClick={jest.fn()}
            hasError={false}
            hasMaxError={false}
            hasMinError={false}
          />
        </IntlProvider>
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
