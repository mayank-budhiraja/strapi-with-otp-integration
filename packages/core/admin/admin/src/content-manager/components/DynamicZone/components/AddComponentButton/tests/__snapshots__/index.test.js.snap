// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddComponentButton /> displays the name of the dz when the label is empty 1`] = `
.c10 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  padding: 8px;
  border-radius: 4px;
  background: #ffffff;
  border: 1px solid #dcdce4;
  position: relative;
  outline: none;
}

.c1 svg {
  height: 12px;
  width: 12px;
}

.c1 svg > g,
.c1 svg path {
  fill: #ffffff;
}

.c1[aria-disabled='true'] {
  pointer-events: none;
}

.c1:after {
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border: 2px solid transparent;
}

.c1:focus-visible {
  outline: none;
}

.c1:focus-visible:after {
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -5px;
  bottom: -5px;
  left: -5px;
  right: -5px;
  border: 2px solid #4945ff;
}

.c4 {
  padding-right: 8px;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c9 {
  font-weight: 600;
  color: #8e8ea9;
  font-size: 0.75rem;
  line-height: 1.33;
}

.c7 {
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c7 > circle {
  fill: #eaeaef;
}

.c7 > path {
  fill: #666687;
}

.c2 {
  border-radius: 26px;
  border-color: #eaeaef;
  background: #ffffff;
  padding-top: 12px;
  padding-right: 16px;
  padding-bottom: 12px;
  padding-left: 16px;
  box-shadow: 0px 1px 4px rgba(33,33,52,0.1);
}

.c2 svg {
  height: 24px;
  width: 24px;
}

.c2 svg > path {
  fill: #666687;
}

.c2:hover {
  color: #4945ff !important;
}

.c2:hover .c8 {
  color: #4945ff !important;
}

.c2:hover .c6 > circle {
  fill: #4945ff;
}

.c2:hover .c6 > path {
  fill: #f6f6f9;
}

.c2:active .c8 {
  color: #4945ff;
}

.c2:active .c6 > circle {
  fill: #4945ff;
}

.c2:active .c6 > path {
  fill: #f6f6f9;
}

.c5 {
  height: 100%;
}

<div>
  <div
    class="c0"
  >
    <div
      class=""
      style="cursor: pointer;"
    >
      <button
        aria-disabled="false"
        class="c1 c2"
        type="button"
      >
        <div
          class="c3"
        >
          <div
            aria-hidden="true"
            class="c4 c5"
          >
            <svg
              class="c6 c7"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                fill="#212134"
                r="12"
              />
              <path
                d="M17 12.569c0 .124-.1.224-.225.224h-3.981v3.982c0 .124-.101.225-.226.225h-1.136a.225.225 0 01-.226-.225v-3.981H7.226A.225.225 0 017 12.567v-1.136c0-.125.1-.226.225-.226h3.982V7.226c0-.124.1-.225.224-.225h1.138c.124 0 .224.1.224.225v3.982h3.982c.124 0 .225.1.225.224v1.138z"
                fill="#F6F6F9"
              />
            </svg>
          </div>
          <span
            class="c8 c9"
          >
            Add a component to name
          </span>
        </div>
      </button>
    </div>
  </div>
  <div
    class="c10"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`<AddComponentButton /> renders and matches the snapshot 1`] = `
.c10 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  padding: 8px;
  border-radius: 4px;
  background: #ffffff;
  border: 1px solid #dcdce4;
  position: relative;
  outline: none;
}

.c1 svg {
  height: 12px;
  width: 12px;
}

.c1 svg > g,
.c1 svg path {
  fill: #ffffff;
}

.c1[aria-disabled='true'] {
  pointer-events: none;
}

.c1:after {
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border: 2px solid transparent;
}

.c1:focus-visible {
  outline: none;
}

.c1:focus-visible:after {
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -5px;
  bottom: -5px;
  left: -5px;
  right: -5px;
  border: 2px solid #4945ff;
}

.c4 {
  padding-right: 8px;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c9 {
  font-weight: 600;
  color: #8e8ea9;
  font-size: 0.75rem;
  line-height: 1.33;
}

.c7 {
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c7 > circle {
  fill: #eaeaef;
}

.c7 > path {
  fill: #666687;
}

.c2 {
  border-radius: 26px;
  border-color: #eaeaef;
  background: #ffffff;
  padding-top: 12px;
  padding-right: 16px;
  padding-bottom: 12px;
  padding-left: 16px;
  box-shadow: 0px 1px 4px rgba(33,33,52,0.1);
}

.c2 svg {
  height: 24px;
  width: 24px;
}

.c2 svg > path {
  fill: #666687;
}

.c2:hover {
  color: #4945ff !important;
}

.c2:hover .c8 {
  color: #4945ff !important;
}

.c2:hover .c6 > circle {
  fill: #4945ff;
}

.c2:hover .c6 > path {
  fill: #f6f6f9;
}

.c2:active .c8 {
  color: #4945ff;
}

.c2:active .c6 > circle {
  fill: #4945ff;
}

.c2:active .c6 > path {
  fill: #f6f6f9;
}

.c5 {
  height: 100%;
}

<div>
  <div
    class="c0"
  >
    <div
      class=""
      style="cursor: pointer;"
    >
      <button
        aria-disabled="false"
        class="c1 c2"
        type="button"
      >
        <div
          class="c3"
        >
          <div
            aria-hidden="true"
            class="c4 c5"
          >
            <svg
              class="c6 c7"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                fill="#212134"
                r="12"
              />
              <path
                d="M17 12.569c0 .124-.1.224-.225.224h-3.981v3.982c0 .124-.101.225-.226.225h-1.136a.225.225 0 01-.226-.225v-3.981H7.226A.225.225 0 017 12.567v-1.136c0-.125.1-.226.225-.226h3.982V7.226c0-.124.1-.225.224-.225h1.138c.124 0 .224.1.224.225v3.982h3.982c.124 0 .225.1.225.224v1.138z"
                fill="#F6F6F9"
              />
            </svg>
          </div>
          <span
            class="c8 c9"
          >
            Add a component to test
          </span>
        </div>
      </button>
    </div>
  </div>
  <div
    class="c10"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;

exports[`<AddComponentButton /> renders and matches the snapshot when the isOpen prop is truthy 1`] = `
.c10 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  padding: 8px;
  border-radius: 4px;
  background: #ffffff;
  border: 1px solid #dcdce4;
  position: relative;
  outline: none;
}

.c1 svg {
  height: 12px;
  width: 12px;
}

.c1 svg > g,
.c1 svg path {
  fill: #ffffff;
}

.c1[aria-disabled='true'] {
  pointer-events: none;
}

.c1:after {
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border: 2px solid transparent;
}

.c1:focus-visible {
  outline: none;
}

.c1:focus-visible:after {
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -5px;
  bottom: -5px;
  left: -5px;
  right: -5px;
  border: 2px solid #4945ff;
}

.c4 {
  padding-right: 8px;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c9 {
  font-weight: 600;
  color: #8e8ea9;
  font-size: 0.75rem;
  line-height: 1.33;
}

.c7 {
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c7 > circle {
  fill: #eaeaef;
}

.c7 > path {
  fill: #666687;
}

.c2 {
  border-radius: 26px;
  border-color: #eaeaef;
  background: #ffffff;
  padding-top: 12px;
  padding-right: 16px;
  padding-bottom: 12px;
  padding-left: 16px;
  box-shadow: 0px 1px 4px rgba(33,33,52,0.1);
}

.c2 svg {
  height: 24px;
  width: 24px;
}

.c2 svg > path {
  fill: #666687;
}

.c2:hover {
  color: #4945ff !important;
}

.c2:hover .c8 {
  color: #4945ff !important;
}

.c2:hover .c6 > circle {
  fill: #4945ff;
}

.c2:hover .c6 > path {
  fill: #f6f6f9;
}

.c2:active .c8 {
  color: #4945ff;
}

.c2:active .c6 > circle {
  fill: #4945ff;
}

.c2:active .c6 > path {
  fill: #f6f6f9;
}

.c5 {
  height: 100%;
}

<div>
  <div
    class="c0"
  >
    <div
      class=""
      style="cursor: pointer;"
    >
      <button
        aria-disabled="false"
        class="c1 c2"
        type="button"
      >
        <div
          class="c3"
        >
          <div
            aria-hidden="true"
            class="c4 c5"
          >
            <svg
              class="c6 c7"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                fill="#212134"
                r="12"
              />
              <path
                d="M17 12.569c0 .124-.1.224-.225.224h-3.981v3.982c0 .124-.101.225-.226.225h-1.136a.225.225 0 01-.226-.225v-3.981H7.226A.225.225 0 017 12.567v-1.136c0-.125.1-.226.225-.226h3.982V7.226c0-.124.1-.225.224-.225h1.138c.124 0 .224.1.224.225v3.982h3.982c.124 0 .225.1.225.224v1.138z"
                fill="#F6F6F9"
              />
            </svg>
          </div>
          <span
            class="c8 c9"
          >
            Close
          </span>
        </div>
      </button>
    </div>
  </div>
  <div
    class="c10"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
